llvm-hs/src/LLVM/Internal/OrcJIT.hs

```diff
@@ -148,3 +149,20 @@ createRegisteredDataLayout (TargetMachine tm) cleanups =
         modifyIORef' cleanups (FFI.disposeDataLayout dl :)
         pure dl
   in anyContToM $ bracketOnError createDataLayout FFI.disposeDataLayout
+
+-- | Add an object file to the 'LinkingLayer'. The 'SymbolResolver' is used
+-- to resolve external symbols in the module.
+addObjectFile :: LinkingLayer l => l -> ObjectFile -> SymbolResolver
+              -> IO FFI.ObjectHandle
+addObjectFile linkingLayer obj resolver = flip runAnyContT return $ do
+  cleanups    <- liftIO $ newIORef ([] :: [IO ()])
```

Comment:
Is this correct in this context? [CompileOnDemandLayer](https://github.com/llvm-hs/llvm-hs/blob/llvm-6/llvm-hs/src/LLVM/Internal/OrcJIT/CompileOnDemandLayer.hs#L41) is initialized like this.

Reply:
If you take a look at `newCompileOnDemandLayer`, it stores the cleanups and `disposeCompileLayer` the runs them. In your implementation you are just throwing away the cleanups which would result in a memory leak. What needs to be done here is to add a method to the [LinkingLayer class](https://github.com/llvm-hs/llvm-hs/blob/66bbb89ddf6aafe4032021d920e60b377db9460a/llvm-hs/src/LLVM/Internal/OrcJIT.hs#L66) to access the cleanups similar to how the corresponding method in [CompileLayer](https://github.com/llvm-hs/llvm-hs/blob/66bbb89ddf6aafe4032021d920e60b377db9460a/llvm-hs/src/LLVM/Internal/OrcJIT/CompileLayer.hs#L30), add a field to `ObjectLinkingLayer` to store the cleanups and modify [disposeLinkingLayer](https://github.com/llvm-hs/llvm-hs/blob/66bbb89ddf6aafe4032021d920e60b377db9460a/llvm-hs/src/LLVM/Internal/OrcJIT.hs#L132) to run the cleanups.

Reply:
Oh no. This was a stupid mistake. Fixing it.